src/analysis/souffle/examples/vision_pipeline/VisionPipelineAuthLogic

```diff
@@ -0,0 +1,96 @@
+        // Ideally, we would check that the value of the 
+        // identified product matches a value produced by the particle,
+        // but this requires some amount of dynamic enforcement.
+
+    // Delegates TPPR the ability to apply tags anywhere
+    "ThirdPartyPrivacyReviewer" canSay hasTag(x, y) :- 
+        isAccessPath(x), isTag(y).
+
+    // Delegate the right to claim that particles are "Local" to TPPR.
+    // (We can think of the principal "LocalDevice" as a group)
+    "ThirdPartyPrivacyReviewer" canSay principalX canActAs "LocalDevice"
+
+    "LocalDevice" may doPureComputation() using "raw_video_tag".
+        // this might ideally be
+        // "LocalDevice" may x() using "product_id_tag" :- x != egressTo
+        // but to express this we would need another language above this one
+    "LocalDevice" may egressTo("ShoppingService") using "product_id_tag".
+
+    // Could also give different permissions to RemoteService for once it
+    // has "product_id_tag".
+}
+
+"ThirdPartyPrivacyReviewer" says {
+    "VideoDataSource" canActAs "LocalDevice".
+    "ImageDetector" canActAs "LocalDevice".
+    "ImageSelector" canActAs "LocalDevice".
+    "ProductIdOutput" canActAs "LocalDevice".
+    
+    // TPPR delegates the data transformation to ImageSelector
+    "ImageSelector" canSay becomes(pathX, "raw_video_tag", "product_id_tag") :-
+        isAccessPath(pathX).
+    "ImageSelector" canSay objectSelected("EndUser", pathX) :-
+        isAccessPath(pathX).
+}
+
+// Maybe we want this part of the policy to sit closer to the particles
+// because sometimes relations will refer to path names ?
+// (We can ostensibly do this as long as we can write
+// `claim <RelationName(...)>` in manifests)
+"VideoDataSource" says will doPureComputation() using "raw_video_tag".
+"ImageDetector" says will doPureComputation() using "raw_video_tag".
+"ImageSelector" says {
+    will doPureComputation() using "raw_video_tag".
+    becomes(pathX, "raw_video_tag", "product_id_tag").
+    objectSelected("EndUser", pathX).
+}
+
+/// Some language features are just "imagined" at the moment. 
+//
+// Becomes:
+// `becomes(path, oldTag, newTag)` which is the same as 
+// declassify(path, oldTag, newTag) and claimHasTag(path, oldTag, newTag)
+// (Maybe a more general way to implement this is macros in auth logic)
+//
+// May / Will:
+// These are used to specify allowed usages and a check that only the allowed 
+// usages are done. The check depends on the taint analysis. Roughly:
+//
+// For each particle P, for each tag t in union over tags in read edges of P,
+//      for each statement of the form:
+//          `P says will <RELATION> using <TAG>`
+//      prove
+//          `ownsTag(Q, <TAG>)`
+//          `Q says may <RELATION> using <TAG>`
+//
+// Notably "may" is a keyword rather than just another relation itself because it 
+// seems useful to let usages be arbitrary relations (with potentially varying 
+// signatures).
```

Comment:
On this `may` and `will` part, I think that `may` is definitely in the scope of the authorization logic, but I'm wondering whether `will` is.

The distinction between `may` and `will` actually sums up the distinction in my mind between the concepts that should go in the Manifest (or what the manifest will become) vs. what goes in the authorization logic. The Manifest is an abstract description of the behaviors that either will definitely happen or will possibly happen. The authorization logic, on the other hand, is the description of the behaviors that are permitted to occur. The point of Raksha is to abstractly execute the abstract description of the program's behavior to see if what can happen can be made to conflict with what may happen.

The abstract description of behavior that you are placing on `will` via arbitrary relations may speak to a desire to have `ParticleSpec`s be more than a black box; perhaps they could be augmented with some abstract description of their internal behavior, which may then be governed via the `may` statements you use here.

What do you think about this distinction?

Reply:
Will we need this distinction after the MVP? I thought we were moving towards having just one language. Once we have our own (one) parser, I think it should be fairly easy to have language constructs from authorization logic in the relevant parts of what is currently expressed with manifests as well. If we do it that way, we can easily use `will ...` in a way that refers to paths in particles accurately (which is what I think you want to gain by putting this functionality in manifests) but would still allow this language feature to interact with principals and other concepts from authorization language in the intended way. This would also allow it to be applied to relations of arbitrary arity (which I think we would need if the `may` construct can be applied to arbitrary relations). 

I don't know a ton about the parser combinators that Gogul mentioned last time, but using one of these seems like it would help us both: 1) use authorization logic constructs in arbitrary places of the raksha syntax, and 2) if we want to, keep an implementation of authorization logic as a separate modular block. (Where doing 2) could be useful because it might be possible to re-use  authorization logic for other use-cases that are not about use of privacy-sensitive data -- then we'll all have two broadly applicable pieces of software that we are all involved in producing. Though this is probably lower priority than getting our one combined thing to work well first).

Reply:
I think for the MVP, there is probably a way to move the `will` parts into the manifests where we can just have some additional kind of annotation that means `ParticleAsPrincipal says ParticleAsPrincipal will Relation`. Exactly the cheapest/easiest way to do this is something Gogul/you might know better. 

Just for the purpose of writing demos while are still working on the MVP, it's useful for me to also be able to write this in auth logic code because these relations interact with other parts of auth logic. If it's useful to be able to write `will()` relations in either place (manifests or auth logic code) we can do the same thing we are doing with `claimHasTag(_, _, _)` and `saysHasTag(_, _, _)` which lets us do it in both places.

Reply:
Even after we have one language for both the authorization and dataflow aspects, it may be useful to keep an eye on how much we entangle them. If we have things separated cleanly enough, it could enable nice properties such as an administrator running multiple experiments with hypothetical policies against a single dataflow graph. That said, the benefits are speculative and I'm not sure such a separation would necessarily work, so let's not worry about it too much right now.

Reply:
If we use the `ParticleAsPrincipal says will <Relation>(<args>...)` syntax within a manifest file, it should be possible to test this one manifest that describes how the application will use data against several distinct policies about allowed data use (the different policies about allowed use could be in separate files from the manifest).

Reply:
We just discussed in our 1:1 and I think we realized we were essentially in agreement that this information should go with specification of the application's behavior. I think we are kind of undecided about what exactly the syntax of allowed usage will be.

(Closing for now, but feel free to re-open)